bigtable/benchmarks/embedded_server.cc

```diff
@@ -0,0 +1,200 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/benchmarks/embedded_server.h"
+
+#include <iomanip>
+#include <sstream>
+
+#include "google/bigtable/v2/bigtable.grpc.pb.h"
+
+#include "bigtable/benchmarks/random.h"
+#include "bigtable/benchmarks/setup.h"
+
+namespace btproto = google::bigtable::v2;
+namespace adminproto = google::bigtable::admin::v2;
+
+namespace bigtable {
+namespace benchmarks {
+/**
+ * Implement the portions of the `google.bigtable.v2.Bigtable` interface
+ * necessary for the benchmarks.
+ *
+ * This is not a Mock (use `google::bigtable::v2::MockBigtableStub` for that,
+ * nor is this a Fake implementation (use the Cloud Bigtable Emulator for that),
+ * this is an implementation of the interface that returns hardcoded values.
+ * It is suitable for the benchmarks, but for nothing else.
+ */
+class BigtableImpl final : public btproto::Bigtable::Service {
+ public:
+  BigtableImpl() {
+    // Prepare a list of random values to use at run-time.  This is because we
+    // want the overhead of this implementation to be as small as possible.
+    // Using a single value is an option, but compresses too well and makes the
+    // tests a bit unrealistic.
+    auto generator = MakeDefaultPRNG();
+    values_.resize(1000);
+    std::generate(values_.begin(), values_.end(),
+                  [&generator]() { return MakeRandomValue(generator); });
+  }
+
+  grpc::Status MutateRow(grpc::ServerContext* context,
+                         btproto::MutateRowRequest const* request,
+                         btproto::MutateRowResponse* response) override {
+    return grpc::Status::OK;
+  }
+
+  grpc::Status MutateRows(
+      grpc::ServerContext* context, btproto::MutateRowsRequest const* request,
+      grpc::ServerWriter<btproto::MutateRowsResponse>* writer) override {
+    btproto::MutateRowsResponse msg;
+    for (int index = 0; index != request->entries_size(); ++index) {
+      auto& entry = *msg.add_entries();
+      entry.set_index(index);
+      entry.mutable_status()->set_code(grpc::StatusCode::OK);
+    }
+    writer->WriteLast(msg, grpc::WriteOptions());
+    return grpc::Status::OK;
+  }
+
+  grpc::Status ReadRows(
+      grpc::ServerContext* context,
+      google::bigtable::v2::ReadRowsRequest const* request,
+      grpc::ServerWriter<google::bigtable::v2::ReadRowsResponse>* writer)
+      override {
+    if (request->rows_limit() == 0) {
+      return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "not-yet");
+    }
+
+    if (request->rows_limit() == 1) {
+      if (request->rows().row_keys_size() != 1) {
+        return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "not-yet");
+      }
+      auto const& row_key = request->rows().row_keys(0);
+      btproto::ReadRowsResponse msg;
+      msg.set_last_scanned_row_key(row_key);
```

Comment:
last scanned row key is only set here for rows_limit == 1. I think it's actually not needed at all

Reply:
Done.